### ISSUE TITLE
Update history table in db to handle system messages, in addition to chat and file messages

### ISSUE THREAD

**Issue by anthonybilinski** [qTox/qTox#6221]
##### Brief Description

"System messages" here is meant to refer to messages that qTox adds to the chatlog, but that weren't sent by either friend. Things like "Bob is calling...", "Call ended after 1h35m", "Bob is now away." etc.

As part of https://github.com/qTox/qTox/projects/5, to allow:
1) A unique reference to any individual chat message in the chatlog. Currently messages are identified in `SessionChatLog` by the `History` index now, but that can't accommodate system messages, which chatlog still needs to handle.
2) Users scrolling to dynamically remove system messages from the chatlog, and also dynamically load them in if they scroll back down, the system messages need to be persisted.
3) To allow UI changes like "Show contacts' status changes" to take immediate effect for the viewed chatlog, by saving system messages even if they aren't displayed.

The history should store system messages, like it does chat and file messages already.

Current the `history` table has all the fields necessary for a chat message, then an optional foreign reference to an id in the `file_transfers` table. Since system messages have quite different requirements from normal chat messages, e.g. no author, and rather than message content just an enum could be saved for which type of system message it was, I think a more flexible design would be:

1. move chat messages to new table, same schema as now. id is a foreign key to history
1. make new system messages table, similar to history but without author
1. update file transfer table to include needed fields that were previously from `history`.
1. make `history` table just an `id` + `type` + `foreign key` that references a row in one of the other 3 tables.

This would allow each of the three sub-history tables to have all of their fields present, and would allow the master history table to still enforce ordering between messages. It should be verified that this wouldn't majorly impact performance and that this is actually a sane schema design.

**Comment by anthonybilinski** [comment]
Researching how to lay this out in SQL a bit, https://stackoverflow.com/a/5471265 and https://stackoverflow.com/a/10077883 look like they do exactly what we want for `history`. Given that, I think our schema should move to this:

```
CREATE TABLE peers (id INTEGER PRIMARY KEY, public_key TEXT NOT NULL UNIQUE);
CREATE TABLE aliases (id INTEGER PRIMARY KEY, owner INTEGER, display_name BLOB NOT NULL, UNIQUE(owner, display_name), FOREIGN KEY (owner) REFERENCES peers(id));
CREATE TABLE history (id INTEGER PRIMARY KEY, message_type char(1) check (message_type in ('T', 'F')), timestamp INTEGER NOT NULL, chat_id INTEGER NOT NULL, FOREIGN KEY (chat_id) REFERENCES peers(id));
CREATE TABLE text_messages (id INTEGER PRIMARY KEY, message_type char(1) not null default 'T' check (message_type = 'T'), sender_alias INTEGER NOT NULL, message BLOB NOT NULL, FOREIGN KEY (id) REFERENCES history(id), FOREIGN KEY (sender_alias) REFERENCES aliases(id));
CREATE TABLE file_transfers (id INTEGER PRIMARY KEY, message_type char(1) not null default 'F' check (message_type = 'F'), sender_alias INTEGER NOT NULL, file_restart_id BLOB NOT NULL, file_name BLOB NOT NULL, file_path BLOB NOT NULL, file_hash BLOB NOT NULL, file_size INTEGER NOT NULL, direction INTEGER NOT NULL, file_state INTEGER NOT NULL, FOREIGN KEY (id) REFERENCES history(id), FOREIGN KEY (sender_alias) REFERENCES aliases(id));
CREATE TABLE faux_offline_pending (id INTEGER PRIMARY KEY, FOREIGN KEY (id) REFERENCES history(id));
CREATE TABLE broken_messages (id INTEGER PRIMARY KEY, FOREIGN KEY (id) REFERENCES history(id));
```

to mirror our current behaviour.

I've kept `faux_offline_pending` and `broken_messages` referencing `history` rather than `text_messages` ids, since both
1. The IDs will be the same anyway because it's a foreign key.
2. Optimistically file transfers will be able to be offline messages and then get broken in the future :P

I've kept common things that I expect to be in any history message, {timestamp, chat_id} in the parent history table.

With this, our `History` class would just need minor changes to select file and text message info from the appropriate tables and change some joins to select the right message_type, as well as more complicated schema upgrade logic. I'll hold off doing this until https://github.com/qTox/qTox/pull/5932 is merged to avoid too many conflicts.

To add system message from there, we just need to add a third `message_type` value of `S` with a new table `system_messages` that has that as the value for every row, and that can have its own schema with no author field :tada:

**Comment by sphaerophoria** [comment]
Take my comment with a grain of salt since my database experience is fairy lacking... but I've looked over the proposed schema and it seems pretty reasonable. Another link that supports the approach https://stackoverflow.com/questions/3579079/how-can-you-represent-inheritance-in-a-database.

The only thing I wasn't really sure about was the usage of the message_type. I think right now it doesn't really provide any extra protection having it in the text_messages and file_transfers database. I think the case this could get messed up is if the program ran the following steps

* Insert item into history with type 'M' (oops we actually got a file transfer and made a mistake)
* Insert item into file_trasnfers with type 'F' as you normally would
* Join tables to read the history back :boom: 

I think we want the protection on insertion if we're going to have the message_type duplicated in the text_messages and file_trasnfers field. Maybe drop a FORIEN KEY (message_type) REFERENCES history(message_type)

**Comment by sphaerophoria** [comment]
Ran a quick test where I altered the message_type to have FORIEGN KEY (id, message_type) REFERENCES history(id, message_type) and it worked as i expected. It didn't let me enter a mismatched key and enforced that the key was the value required for the table, so I think it does what we expect and enforces consistency between the tables.

**Comment by sphaerophoria** [comment]
Took a stab at implementing part of this today. Implemented on top of #5932 since @anthonybilinski posted above that we want to wait for the history changes in that PR.

After a bit of messing around I think I got a schema that works well and enforces our constraints.

```
CREATE TABLE peers (id INTEGER PRIMARY KEY, public_key TEXT NOT NULL UNIQUE)
CREATE TABLE aliases (id INTEGER PRIMARY KEY, owner INTEGER, display_name BLOB NOT NULL, UNIQUE(owner, display_name), FOREIGN KEY (owner) REFERENCES peers(id));
CREATE TABLE history (id INTEGER PRIMARY KEY, message_type CHAR(1) NOT NULL DEFAULT 'T' CHECK (message_type in ('T','F')), timestamp INTEGER NOT NULL, chat_id INTEGER NOT NULL, UNIQUE (id, message_type), FOREIGN KEY (chat_id) REFERENCES peers(id))
CREATE TABLE text_messages (id INTEGER PRIMARY KEY, message_type CHAR(1) NOT NULL CHECK (message_type = 'T'), sender_alias INTEGER NOT NULL, message BLOB NOT NULL, FOREIGN KEY (id, message_type) REFERENCES history(id, message_type), FOREIGN KEY (sender_alias) REFERENCES aliases(id))
CREATE TABLE file_transfers (id INTEGER PRIMARY KEY, message_type CHAR(1) NOT NULL CHECK (message_type = 'F'), sender_alias INTEGER NOT NULL, file_restart_id BLOB NOT NULL, file_name BLOB NOT NULL, file_path BLOB NOT NULL, file_hash BLOB NOT NULL, file_size INTEGER NOT NULL, direction INTEGER NOT NULL, file_state INTEGER NOT NULL, FOREIGN KEY (id, message_type) REFERENCES history(id, message_type), FOREIGN KEY (sender_alias) REFERENCES aliases(id))
CREATE TABLE faux_offline_pending (id INTEGER PRIMARY KEY, required_extensions INTEGER NOT NULL DEFAULT 0, FOREIGN KEY (id) REFERENCES history(id));
CREATE TABLE broken_messages (id INTEGER PRIMARY KEY, reason INTEGER NOT NULL DEFAULT 0, FOREIGN KEY (id) REFERENCES history(id))
```

I had to make a few tweaks from the original schema proposed by @anthonybilinski.

1. history now has a UNIQUE constraint on (id, message_type)
2. file_transfers and text_messages now rely on the FOREIGN KEY (id, message_type)

These changes were made to ensure that we cannot accidentally insert data into the wrong table. I created a fresh DB and tried the following to ensure that the constraints are being held

```
INSERT INTO peers(id, public_key) VALUES (1, "0000000000000");
INSERT INTO aliases (id, owner, display_name) VALUES (1, 1, "test");
INSERT INTO history(id, message_type, timestamp, chat_id) VALUES (1, 'T', 0, 1);
INSERT INTO text_messages(id, message_type, sender_alias, message) VALUES (1, 'F', 1, "test"); // FAILED (bad message type)
INSERT INTO text_messages(id, message_type, sender_alias, message) VALUES (1, 'T', 1, "test"); // Success!
INSERT INTO history(id, message_type, timestamp, chat_id) VALUES (2, 'T', 0, 1);
INSERT INTO text_messages(id, message_type, sender_alias, message) VALUES (2, 'T', 2, "test"); // FAILED (bad sender alias)
INSERT INTO history(id, message_type, timestamp, chat_id) VALUES (3, 'F', 0, 1);
INSERT INTO file_transfers(id, message_type, sender_alias, file_restart_id, file_name, file_path, file_hash, file_size, direction, file_state) VALUES (3, 'T', 1, "asdf", "asdf", "asdf", "asdf", 0, 1, 1); // FAILED (bad message type)
INSERT INTO file_transfers(id, message_type, sender_alias, file_restart_id, file_name, file_path, file_hash, file_size, direction, file_state) VALUES (3, 'F', 1, "asdf", "asdf", "asdf", "asdf", 0, 1, 1); // Success!
INSERT INTO history(id, message_type, timestamp, chat_id) VALUES (4, 'F', 0, 1);
INSERT INTO file_transfers(id, message_type, sender_alias, file_restart_id, file_name, file_path, file_hash, file_size, direction, file_state) VALUES (4, 'F', 2, "asdf", "asdf", "asdf", "asdf", 0, 1, 1); // FAILED (bad sender alias)
INSERT INTO history(id, message_type, timestamp, chat_id) VALUES (5, 'F', 0, 1);
INSERT INTO text_messages(id, message_type, sender_alias, message) VALUES (5, 'T', 1, "test"); // FAILED (message type does not match)
INSERT INTO text_messages(id, message_type, sender_alias, message) VALUES (5, 'F', 1, "test"); // FAILED (message type is not 'T')
INSERT INTO file_transfers(id, message_type, sender_alias, file_restart_id, file_name, file_path, file_hash, file_size, direction, file_state) VALUES (5, 'F', 1, "asdf", "asdf", "asdf", "asdf", 0, 1, 1); // Success!
INSERT INTO history(id, message_type, timestamp, chat_id) VALUES (6, 'T', 0, 1);
INSERT INTO file_transfers(id, message_type, sender_alias, file_restart_id, file_name, file_path, file_hash, file_size, direction, file_state) VALUES (6, 'F', 1, "asdf", "asdf", "asdf", "asdf", 0, 1, 1); // FAILED (message type does not match)
INSERT INTO file_transfers(id, message_type, sender_alias, file_restart_id, file_name, file_path, file_hash, file_size, direction, file_state) VALUES (6, 'T', 1, "asdf", "asdf", "asdf", "asdf", 0, 1, 1); // FAILED (message type is not 'F')
```

What I'm unsure of is if the new UNIQUE constraint is going to have implications on performance of history. I would assume no since id already has to be unique so UNIQUE(id, message_type) shouldn't have to do anything to enforce that constraint, but maybe it isn't that smart. We'll have to monitor performance on insertions to see if this is noticeably bad with a big db

**Comment by sphaerophoria** [comment]
Alright I have something that works as far as I can tell. I need to do a little more testing, and maybe should do a little more optimization to remove some of the uses of `generatePeerIdString()` on queries, but I'll leave that for another day. The current work is sitting on top of #5932 and #6288. I'll hold this PR hostage until we can resolve the remaining work required to get those PRs in :P
